api/server/errorhandler.go

```diff
@@ -17,18 +22,99 @@ func makeErrorHandler(err error) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
 		statusCode := httpstatus.FromError(err)
 		vars := mux.Vars(r)
-		if apiVersionSupportsJSONErrors(vars["version"]) {
-			response := &types.ErrorResponse{
-				Message: err.Error(),
-			}
-			_ = httputils.WriteJSON(w, statusCode, response)
-		} else {
+		version := vars["version"]
+
+		const firstAPIVersionWithJSONErrors = "1.24"
+		if version != "" && versions.LessThan(version, firstAPIVersionWithJSONErrors) {
 			http.Error(w, status.Convert(err).Message(), statusCode)
+			return
+		}
+
+		response := marshalErrorResponse(err)
+		const firstAPIVersionWithJSONMultiErrors = "1.44"
+		if version != "" && versions.LessThan(version, firstAPIVersionWithJSONMultiErrors) {
+			response.Errors = nil
+		}
+
+		_ = httputils.WriteJSON(w, statusCode, response)
+	}
+}
+
+type joiningError interface {
+	error
+	Unwrap() []error
+}
+
+type wrappingError interface {
+	error
+	Unwrap() error
+}
+
+// marshalErrorResponse returns a tree of [*types.ErrorResponse] representing the argument err.
+func marshalErrorResponse(err error) *types.ErrorResponse {
+	if errdefs.IsHTTPOnlyError(err) {
+		if wrappedErr := errors.Unwrap(err); wrappedErr != nil {
+			err = wrappedErr
+		}
+	}
+
+	switch err := err.(type) {
+	case joiningError:
+		return marshalJoinedErrors(err)
+	case *multierror.Error:
+		return marshalMultiErrors(err)
+	case wrappingError:
+		return &types.ErrorResponse{
+			Message: err.Error(),
+			Errors:  []*types.ErrorResponse{marshalErrorResponse(err.Unwrap())},
```

Comment:
I'm wondering if we should unconditionally propagate the `Errors` array; my thinking here is that the top-level `Message` can be considered a "pre-rendered" presentation of the error (or _errors_ if there's more than one), which can be used by clients to present as-is, and the `Errors` _slice_ contains the raw, underlying error(s).


Clients that support the new `Errors` field would look in the `Errors` slice, and use it (to present to their liking), and otherwise fall back to using the top-level `Message`  "as-is".

Reply:
That's what this code is doing, except for the limiting case where there are no underlying errors. Note the recursion with `Errors`: it's the same type all the way down. Any property which could go on an object inside `Errors` could also go on the root `ErrorResponse` for a knowledgeable client to use to their liking. Take a look at the details block in my earlier non-inline review comment for a worked example of what that hypothetically might look like.